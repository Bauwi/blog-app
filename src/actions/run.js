import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
