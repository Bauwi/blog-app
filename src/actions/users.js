import { db } from '../firebase/firebase';

export const updateUser = (id, updates) => ({
  type: 'UPDATE_USER',
  id,
  updates
});

export const startUpdateUser = (id, updates) => (dispatch, getState) => {
  const uid = id || getState().auth.uid;
  return db
    .collection('users')
    .doc(uid)
    .get()
    .then((doc) => {
      if (doc.exists) {
        return db
          .collection('users')
          .doc(uid)
          .update(updates)
          .then(() => {
            dispatch(updateUser(uid, updates));
          });
      }
      return db
        .collection('users')
        .doc(uid)
        .set({
          avatar: 'http://www.perlworkshop.nl/nlpw2016/img/default-avatar.png',
          email: '',
          username: '',
          stars: 0,
          topCategories: ['life', 'programming', 'Music']
        })
        .then(startUpdateUser(uid, updates));
    });
};

export const addUserStar = nextStars => ({
  type: 'ADD_USER_STAR',
  nextStars
});

export const startAddUserStar = (id, prevStars) => dispatch =>
  db
    .collection('users')
    .doc(id)
    .update({ stars: prevStars + 1 })
    .then(() => dispatch(addUserStar(prevStars + 1)));

export const setAuthor = author => ({
  type: 'SET_AUTHOR',
  author
});

export const startSetAuthor = id => dispatch =>
  db
    .collection('users')
    .doc(id)
    .get()
    .then((doc) => {
      dispatch(setAuthor(doc.data()));
    });

export const setUserPreferences = preferences => ({
  type: 'SET_USER_PREFERENCES',
  preferences
});

export const startSetUserPreferences = preferences => (dispatch, getState) =>
  db
    .collection('users')
    .doc(getState().auth.uid)
    .get()
    .then((doc) => {
      dispatch(setUserPreferences(doc.data()));
    });
