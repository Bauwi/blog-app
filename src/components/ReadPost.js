import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import Header from './Header';

export const ReadPost = ({
  history, post: {
    id, author, title, body, createdAt, keywords
  }
}) => {
  const regEx = new RegExp(id, 'g');
  const goBackPath = history.location.pathname.replace(regEx, '');
  return (
    <div>
      <Header />
      <header>
        <button onClick={() => history.push(goBackPath)}>Back</button>
        <h2>{title}</h2>
        <p>{createdAt}</p>
        <p>{author}</p>
      </header>
      <section dangerouslySetInnerHTML={{ __html: body }} />
      <footer>keywords: {keywords}</footer>
      <p>Read unique Post</p>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  post: state.posts.find(post => post.id === props.match.params.id)
});

export default withRouter(connect(mapStateToProps)(ReadPost));
