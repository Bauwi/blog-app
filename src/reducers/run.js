const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
