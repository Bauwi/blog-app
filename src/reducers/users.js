const usersReducerDefaultState = [];

const usersReducer = (state = usersReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_USER':
      return {
        ...state,
        preferences: action.newUser
      };
    case 'UPDATE_USER_PREFERENCES':
      return {
        ...state,
        preferences: { ...state.preferences, ...action.updates }
      };
    case 'SET_AUTHOR':
      return {
        ...state,
        author: action.author
      };
    case 'ADD_USER_STAR':
      return {
        ...state,
        author: {
          ...state.author,
          stars: action.nextStars
        }
      };
    case 'SET_USER_PREFERENCES':
      return {
        ...state,
        preferences: action.preferences
      };
    case 'USERS_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'USERS_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    default:
      return state;
  }
};

export default usersReducer;
