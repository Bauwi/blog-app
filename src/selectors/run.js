const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
