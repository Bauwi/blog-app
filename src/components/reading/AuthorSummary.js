import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import UserCard from '../UserCard';
import CompleteFiltersBar from '../filters/CompleteFiltersBar';
import AuthorSummaryList from './AuthorSummaryList';

import selectPosts from '../../selectors/posts';
import LoadingPage from '../LoadingPage';

import { startSetSpecificUserPosts } from '../../actions/readings';

export class AuthorSummary extends Component {
  componentDidMount() {
    this.props.startSetSpecificUserPosts(this.props.userId);
  }

  render() {
    if (this.props.isLoading) {
      return <LoadingPage />;
    }
    return (
      <div>
        <div className="page-container">
          <div className="content-container">
            <UserCard author={this.props.author} userId={this.props.userId} />
            <CompleteFiltersBar />
          </div>
          <AuthorSummaryList posts={this.props.posts} />
        </div>
      </div>
    );
  }
}

AuthorSummary.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  author: PropTypes.object,
  userId: PropTypes.string.isRequired,
  isLoading: PropTypes.bool,
  startSetSpecificUserPosts: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startSetSpecificUserPosts: id => dispatch(startSetSpecificUserPosts(id))
});

const mapStateToProps = (state, props) => ({
  posts: selectPosts(state.readings.currentUserPosts, state.filters),
  author: state.users.author,
  isLoading: state.readings.isLoading,
  userId: props.location.pathname.slice(1, 29)
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthorSummary);
