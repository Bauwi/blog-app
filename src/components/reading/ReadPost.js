import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import ReactQuill from 'react-quill';
import moment from 'moment';

import UserCard from '../UserCard';
import LoadingPage from '../LoadingPage';
import KeywordsList from '../KeywordsList';

import { startSetOnePost } from '../../actions/readings';
import { startUpPostStar } from '../../actions/posts';

export class ReadPost extends Component {
  static defaultProps = {
    post: {
      title: '',
      cover: '',
      body: [],
      keywords: ''
    }
  };
  componentDidMount() {
    this.props.startSetOnePost(this.props.postId);
  }

  onAddStar = () => {
    const { postId } = this.props;
    const { authorId } = this.props.post;
    const { stars } = this.props.author;
    this.props.startUpPostStar(postId, authorId, stars);
  };

  render() {
    if (this.props.isLoading) {
      return <LoadingPage />;
    }
    const { title, cover, body, keywords, stars, readingTime, createdAt } = this.props.post;
    const category = keywords.split(',')[0].trim();
    return (
      <div>
        <div>
          <header className="page-header page-header--read">
            <div className="content-container">
              <UserCard post={this.props.post} author={this.props.author} />
              <div className={`read-header__datereadingtime border-top-item-${category}`}>
                <p>{moment(createdAt).format('MMM Do, YYYY')}</p>
                <p>
                  <i className="fa fa-clock-o" />{' '}
                  {`${readingTime} ${readingTime >= 2 ? 'minutes' : 'minute'} read`}
                </p>
                <div className="read-header__stars">
                  <button className="button button--star" onClick={this.onAddStar}>
                    {stars}
                    <i className="fa fa-star-o" />
                  </button>
                  <p>{this.props.postStars}</p>
                </div>
              </div>
            </div>
          </header>
          <h2 className="read-header__title">{title}</h2>

          <img className="image image--cover" src={cover} />
          <div className="content-container read-only">
            <ReactQuill
              theme="snow"
              value={{ ops: [...body] }}
              readOnly
              modules={ReadPost.modules}
              bounds=".app"
              placeholder="Add a post to your blog"
            />
          </div>

          <footer className={`keywords-list__read-post-container border-top-item-${category}`}>
            <KeywordsList keywords={keywords} />
          </footer>
        </div>
      </div>
    );
  }
}

ReadPost.propTypes = {
  post: PropTypes.object,
  author: PropTypes.object,
  postId: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
  startSetOnePost: PropTypes.func.isRequired,
  startUpPostStar: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => ({
  startSetOnePost: id => dispatch(startSetOnePost(id)),
  startUpPostStar: (id, authorId, stars) => dispatch(startUpPostStar(id, authorId, stars))
});

const mapStateToProps = (state, props) => ({
  post: state.readings.current,
  postId: props.match.params.id,
  author: state.users.author,
  isLoading: state.readings.isLoading
});

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ReadPost));

ReadPost.modules = {
  toolbar: []
};
