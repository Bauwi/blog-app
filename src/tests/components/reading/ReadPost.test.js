import React from 'react';
import { shallow } from 'enzyme';
import { ReadPost } from '../../../components/reading/ReadPost';
import run from '../../fixtures/run';
import { user2 } from '../../fixtures/users';

const posts = run.posts.map(post => post.content);

test('should render ReadPost component properly', () => {
  const wrapper = shallow(<ReadPost
    isLoading={false}
    post={posts[2]}
    postId="1"
    startUpPostStar={() => {}}
    startSetOnePost={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ReadPost properly on loading', () => {
  const wrapper = shallow(<ReadPost
    isLoading
    post={posts[2]}
    postId="1"
    startUpPostStar={() => {}}
    startSetOnePost={() => {}}
  />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startSetOnePost on mount', () => {
  const startSetOnePost = jest.fn();
  const wrapper = shallow(<ReadPost
    isLoading={false}
    post={posts[2]}
    postId="1"
    startUpPostStar={() => {}}
    startSetOnePost={startSetOnePost}
  />);
  expect(startSetOnePost).toHaveBeenLastCalledWith('1');
});

test('should call startUpPostStar on star click', () => {
  const startUpPostStar = jest.fn();
  const wrapper = shallow(<ReadPost
    isLoading={false}
    post={posts[2]}
    author={user2}
    postId="1"
    startUpPostStar={startUpPostStar}
    startSetOnePost={() => {}}
  />);
  wrapper.find('button').simulate('click');
  expect(startUpPostStar).toHaveBeenLastCalledWith('1', posts[2].authorId, user2.stars);
});
