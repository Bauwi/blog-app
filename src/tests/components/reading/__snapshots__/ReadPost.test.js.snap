// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReadPost component properly 1`] = `
<div>
  <div>
    <header
      className="page-header page-header--read"
    >
      <div
        className="content-container"
      >
        <UserCard
          author={
            Object {
              "avatar": "https://www.timeshighereducation.com/sites/default/files/byline_photos/default-avatar.png",
              "description": "I had three chairs in my house; one for solitude, two for friendship, three for society.",
              "stars": 0,
              "username": "Anonymous",
            }
          }
          post={
            Object {
              "author": "author3",
              "authorId": "author3Id",
              "body": Array [
                Object {
                  "attributes": "h3",
                  "text": "blabla",
                },
              ],
              "cover": "https://www.w3schools.com/w3css/img_fjords.jpg",
              "createdAt": 1000000,
              "id": "3",
              "keywords": "car, pool, cake",
              "miniCover": "https://www.w3schools.com/w3css/img_fjords.jpg",
              "readingTime": 2,
              "stars": 2,
              "title": "Post 3 in Run",
            }
          }
        />
        <div
          className="read-header__stars"
        >
          <button
            className="button button--star"
            onClick={[Function]}
          >
            <i
              className="fa fa-star-o"
            />
          </button>
          <p />
        </div>
      </div>
    </header>
    <h2
      className="read-header__title"
    >
      Post 3 in Run
    </h2>
    <img
      className="image image--cover"
      src="https://www.w3schools.com/w3css/img_fjords.jpg"
    />
    <div
      className="content-container read-only"
    >
      <Quill
        bounds=".app"
        modules={
          Object {
            "toolbar": Array [],
          }
        }
        placeholder="Add a post to your blog"
        readOnly={true}
        theme="snow"
        value={
          Object {
            "ops": Array [
              Object {
                "attributes": "h3",
                "text": "blabla",
              },
            ],
          }
        }
      />
    </div>
    <footer
      className="keywords-list__read-post-container "
    >
      <KeywordsList
        keywords="car, pool, cake"
      />
    </footer>
  </div>
</div>
`;

exports[`should render ReadPost properly on loading 1`] = `<Component />`;
